### DISCUSSION TITLE
can i use softdeletes in adonis v5

### DISCUSSION THREAD

**Discussion by kmihiranga** [adonisjs/core#1791] [Help]
i have used Laravel many years and i used soft deletes to my many projects. can i know if there any way to use like this soft deletes in Adonis js. i read the documentation but i couldn't get any point using this feature.

**Comment by Obapelumi** [comment]
It's not a built-in feature in Adonisjs yet but you can use the [hooks API](https://preview.adonisjs.com/guides/models/hooks) to achieve this. 
For example, in your migration, you add a `deleted_at` column:
```
    this.schema.createTable(this.tableName, (table) => {
      //...
      table.dateTime("deleted_at").nullable();
    });
```

Then in your model you add @beforeDelete, @beforeFind and @beforeFetch hooks lie so:

```
export default class User extends BaseModel {
  @beforeDelete()
  public static async softDelete (user: User) {
    user.deletedAt = DateTime.local().toSQL()
    await user.save()
  }

  @beforeFind()
  public static ignoreDeleted (query: ModelQueryBuilderContract<typeof User>) {
    query.whereNull('deleted_at')
  }

  @beforeFetch()
  public static ignoreDeleted (query: ModelQueryBuilderContract<typeof User>) {
    query.whereNull('deleted_at')
  }
}
```

I am not entirely sure about the behaviour of the @beforeDelete hook though. An alternative would be to override the default delete method in your model. for example:

```
  public async delete() {
    this.deletedAt = DateTime.local().toSQL()
    await this.save()
  }
```

**Reply by Obapelumi** [reply]
In my project, I've actually abstracted these out so they are re-usable. For example: 
```
export const softDelete = async (row: LucidRow) => {
  row["deletedAt"] = DateTime.local().toSQL();
  await row.save();
};

export const softDeleteQuery = (
  query: ModelQueryBuilderContract<typeof BaseModel>
) => query.whereNull("deleted_at");
```

Then on the model:

```
export default class User extends BaseModel {
  @beforeDelete()
  public static softDelete = softDelete;

  @beforeFind()
  public static softDeletesFind = softDeleteQuery;

  @beforeFetch()
  public static softDeletesFetch = softDeleteQuery;
}
```

**Reply by kavience** [reply]
but it will make a error while the model has relationships, errors like this:

      "sqlMessage": "Column 'deletedAt' in where clause is ambiguous"

**Reply by Obapelumi** [reply]
Yes! Happens if you are querying from more than one table. You'll need to add the table name. To do that you can replace:

```
export const softDeleteQuery = (
  query: ModelQueryBuilderContract<typeof BaseModel>
) => query.whereNull("deleted_at");
```
with:

```
export const softDeleteQuery = (
  query: ModelQueryBuilderContract<typeof BaseModel>
) => {
  const column = `${query.model.table}.deleted_at`;
  query.whereNull(column);
}

```

**Reply by kavience** [reply]
Yes, I found it from https://github.com/lookinlab/adonis-lucid-soft-deletes

**Reply by darshithedpara** [reply]
[https://github.com/lookinlab/adonis-lucid-soft-deletes](https://github.com/lookinlab/adonis-lucid-soft-deletes) Package is good but it works only with signal records when you try to mass delete 

`await Categories.query().whereIn('id', idsArray).delete()`

it performs the hard delete.

See issue : [https://github.com/lookinlab/adonis-lucid-soft-deletes/issues/9](https://github.com/lookinlab/adonis-lucid-soft-deletes/issues/9)

**Comment by varun-nambiar** [comment]
@kmihiranga where are you on this now?

Is soft delete a planned feature in adonis5?

**Comment by kavience** [comment]
https://github.com/adonisjs/core/issues/3411

Finally, I solved the problem temporarily in this way:

define Base in Models `app/Models/Base.ts`, and all models need soft delete extends Base.
```ts
import { DateTime } from 'luxon'
import { BaseModel, beforeFetch, beforeFind, column } from '@ioc:Adonis/Lucid/Orm'
import CamelCaseNamingStrategy from 'App/Helper/CamelCaseNamingStrategy'
import { softDelete, softDeleteQuery } from 'App/Services/SoftDelete'

export default class Base extends BaseModel {
  public static namingStrategy = new CamelCaseNamingStrategy() // this maybe not needed for you

  @column.dateTime({ serializeAs: null })
  public deletedAt: DateTime

  @beforeFind()
  public static softDeletesFind = softDeleteQuery

  @beforeFetch()
  public static softDeletesFetch = softDeleteQuery

  public async softDelete(column?: string) {
    await softDelete(this, column)
  }
}
```

app/Services/SoftDelete.ts
```ts
import { BaseModel, LucidRow, ModelQueryBuilderContract } from '@ioc:Adonis/Lucid/Orm'
import { DateTime } from 'luxon'

export const softDeleteQuery = (query: ModelQueryBuilderContract<typeof BaseModel>) => {
  query.whereNull(`${query.model.table}.deletedAt`)
}

export const softDelete = async (row: LucidRow, column: string = 'deletedAt') => {
  row[column] = DateTime.local()
  await row.save()
  return
}
```

hope can help someone.
